stdlib/pagerduty/pagerduty.go

```diff
@@ -0,0 +1,188 @@
+	if _, ok := qs.(*DedupKeyOpSpec); !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+	return &DedupProcedureSpec{}, nil
+}
+
+type DedupKeyTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+}
+
+func createDedupKeyTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	dataset := execute.NewDataset(id, mode, cache)
+	if _, ok := spec.(*DedupProcedureSpec); !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", spec)
+	}
+
+	transform := NewDedupKeyTransformation(dataset, cache)
+	return transform, dataset, nil
+}
+
+func NewDedupKeyTransformation(d execute.Dataset, cache execute.TableBuilderCache) *DedupKeyTransformation {
+	return &DedupKeyTransformation{
+		d:     d,
+		cache: cache,
+	}
+}
+
+func (t *DedupKeyTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+type kvs struct {
+	k string
+	v string
+}
+
+func (t *DedupKeyTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	groupCols := tbl.Key().Cols()
+	groupVals := tbl.Key().Values()
+	keys := make([]kvs, len(groupCols))
+	for i := range groupCols {
+		keys[i].k = groupCols[i].Label
+		v := groupVals[i]
+		var str string
+		switch v.Type().Nature() {
+		case semantic.String:
+			str = v.Str()
+		case semantic.Int:
+			str = strconv.FormatInt(v.Int(), 10)
+		case semantic.UInt:
+			str = strconv.FormatUint(v.UInt(), 10)
+		case semantic.Float:
+			str = strconv.FormatFloat(v.Float(), 'f', -1, 64)
+		case semantic.Bool:
+			str = strconv.FormatBool(v.Bool())
+		case semantic.Time:
+			str = v.Time().String()
+		case semantic.Duration:
+			str = v.Duration().String()
+		default:
+			return fmt.Errorf("cannot convert %v to string", v.Type())
+		}
+
+		keys[i].v = str
+
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i].k < keys[j].k
+	})
+
+	sb := strings.Builder{}
+	for i := range keys {
+		sb.WriteString(keys[i].k)
+		sb.WriteRune('\n')
+		sb.WriteString(keys[i].v)
+		sb.WriteRune('\n')
+	}
+
+	dedupKey := sb.String()
+
+	builder, isNew := t.cache.TableBuilder(tbl.Key())
+
+	if isNew {
+		if err := execute.AddTableCols(tbl, builder); err != nil { // adds the other columns to builder
+			return err
+		}
+	}
+
+	colIDX, err := builder.AddCol(flux.ColMeta{
+		Label: dedupKeyColName,
+		Type:  flux.TString,
+	})
+
+	// because pagerduty restricts the dedupKey size we are hashing to reduce chance of collisions, by distributing the key more evenly
+	dedupKeyHash := sha256.Sum256([]byte(dedupKey))
+
+	err = tbl.Do(func(cr flux.ColReader) error {
+		l := cr.Len()
+		for j := range builder.Cols() {
+			if j == colIDX {
+				for i := 0; i < l; i++ {
+					if err := builder.AppendValue(j, values.NewString(hex.Dump(dedupKeyHash[:]))); err != nil {
```

Comment:
As a small optimization can you move this part `values.NewString(hex.Dump(dedupKeyHash[:]))`  outside of the `tbl.Do` loop as it doesn't change and is allocation heavy. 

Then you can simply pass the same value in to each row. Values are immutable so you don't have to worry about sharing it.